fabric-dimensions-v1/src/main/java/net/fabricmc/fabric/impl/dimension/FailSoftMapCodec.java

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.impl.dimension;
+
+import java.util.Map;
+import java.util.Objects;
+
+import com.google.common.collect.ImmutableMap;
+import com.mojang.datafixers.util.Pair;
+import com.mojang.serialization.Codec;
+import com.mojang.serialization.DataResult;
+import com.mojang.serialization.DynamicOps;
+import com.mojang.serialization.Lifecycle;
+import com.mojang.serialization.MapLike;
+import com.mojang.serialization.codecs.BaseMapCodec;
+import com.mojang.serialization.codecs.UnboundedMapCodec;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Has the same functionality as {@link UnboundedMapCodec}.
+ * But it will fail-soft when an entry cannot be deserialized.
+ */
+public record FailSoftMapCodec<K, V>(Codec<K> keyCodec, Codec<V> elementCodec) implements BaseMapCodec<K, V>, Codec<Map<K, V>> {
+	private static final Logger LOGGER = LoggerFactory.getLogger("FailSoftMapCodec");
+
+	@Override
+	public <T> DataResult<Pair<Map<K, V>, T>> decode(final DynamicOps<T> ops, final T input) {
+		return ops.getMap(input).setLifecycle(Lifecycle.stable()).flatMap(map -> decode(ops, map)).map(r -> Pair.of(r, input));
+	}
+
+	@Override
+	public <T> DataResult<T> encode(final Map<K, V> input, final DynamicOps<T> ops, final T prefix) {
+		return encode(input, ops, ops.mapBuilder()).build(prefix);
+	}
+
+	/**
+	 * In {@link BaseMapCodec#decode(DynamicOps, MapLike)},
+	 * the whole deserialization will fail if one element fails.
+	 * `apply2stable` will return fail when any of the two elements is failed.
+	 * In this implementation, if one deserialization fails, it will log and ignore.
+	 */
+	@Override
+	public <T> DataResult<Map<K, V>> decode(final DynamicOps<T> ops, final MapLike<T> input) {
```

Comment:
How and why is this different from https://github.com/Mojang/DataFixerUpper/pull/55?

Reply:
I didn't see that PR. Yes this PR probably does similar thing. (The relevant datafixerupper code contains a lot functional abstractions so I am not sure whether that PR works the same as this PR)

Reply:
I see the difference. In that PR, when one deserialization fails, the result will be in "partial result". In this PR, the result will always be normal result.
The `DataResult` can be in 3 states:
* successful with one result
* failed with one partial result
* failed with no partial result
I am not sure how vanilla handles the partial result (DFU is very convoluted). As I tested, always yielding successful result works.

Reply:
Update: In Minecraft 1.19.3, it rejects partial result of dimension data. (`LevelStorage.createLevelDataParser` uses `getOrThrow` with `allowPartial` argument false.) So that PR won't directly solve the issue without extra changes.